### PR TITLE
Introduce clang-format as part of CI using Github Actions

### PR DESCRIPTION
**Pull request by Jakuje**
Testing of #1641

### CODE REVIEW

**Comment by frankmorgner** [comment]
The build fails with missing api_token. I believe I've created GH_TOKEN with the access rights for pushing to OpenSC/Nightly only. Maybe you need to create an other one for pushing to `master`

**Comment by Jakuje** [comment]
Thanks for pointer. I am testing the actions in my fork. Now, I generated my own GH_TOKEN in the fork's PR so lets see how it will work. I will probably push some testing code to be able to check how does it actually work with real code.

https://github.com/Jakuje/OpenSC/pull/1

**Comment by Jakuje** [comment]
@frankmorgner the above is already tested with one of the commits from my Fuzz branch, successfully reporting some diffs through github actions (as I am not that precise with formatting ;) ). The travis does not work because there is probably some old version of clang in the osx builder (does not know `AlignConsecutiveMacros`). I tried with xcode10 but it failed to build altogether so I would probably not mind going only with the github actions

**Comment by frankmorgner** [comment]
I think we're good for merging, right?

**Comment by Jakuje** [comment]
Not yet. I wanted to review some of more obscure options of the code formatter, but did not get to do that yet. Please, keep this one open before we will finalize the format in #1641.

**Comment by frankmorgner** [comment]
What's the status of this pull request?

**Comment by Jakuje** [comment]
> What's the status of this pull request?

It is awfully outdated. But I did not get much feedback on my comments in #1641 during last year so I did not push that, especially when we were in need of release. It is now out so we can try to focus to this agian. I saw @dengert to test this, coming to some issues, but already using something in his PRs.

I will get it up to date and probably try the Github action version as the travis is getting awfully slow recently.

**Comment by Jakuje** [comment]
The initial push I did triggered the diff run against some code half a year old in my master and surprisingly, there were not many issues:

https://github.com/Jakuje/OpenSC/runs/1550132174?check_suite_focus=true

I think most of the issues stem from the places where we did not manage to agree on the preferred approach:
 * maximum line length: @martinpaljak defended 140 lines, @dengert suggested 110 (with good reasoning for code review here), which I second. Third option is the use of 0, which should allow code reflow, but for some reason it does not look like always working.
 * The placement of opening bracket of a function, most of the code has `Attach` value keeping it on the same line as a function, but I would like to propose `AfterFunction` which I find more readable (but do not insist on this).

In some places, I was already rewriting some of the old code to match something like the above formatting as it is pain to read so we already have couple of "reformat" commits in the history and couple of them are in this PR as . I personally do not see it as a big issue and we can consider doing the large reformat commit later after we will resolve the issues. For now, I would like to do the following:

 * Give some time for @OpenSC/maintainers to review this and raise concerns (a month?)
 * Address concerns and merge this GH Action so all new code follows this formatting
 * Consider doing the huge reformat with implementation of `.ignoreRevsFile` (this might make sense to do with `ColumnLimit: 0` to avoid code reflow when not necessary)

**Comment by dengert** [comment]
I have only done: 
- add `// clang-format off' and `// clang-format in `piv-card.c` and `pkcs15-piv.c` to keep some tables readable.
- experimented with vi (.vimrc) https://clang.llvm.org/docs/ClangFormat.html. Search for other editors too. 
- looked at `git blame   --ignore-revs-file xxx` and `git config blame.ignoreRevsFile`

I think it is important that we not lose a lot of the git history if we do any kind of mass reformat of OpenSC. I find git blame to very useful tool.  We can preserve the history if we do thing in an orderly fashion.

1. Keep reformatting changes in their own commits. 

2. Prepare the code before any reformat by protecting tables or structures that have been hand crafted for readability. Collect these changes into a few commits which can then be added to a `.ignoreRevsFile` file at the top lever in the source and commit `.ignoreRevsFile`

3. Recommend developers to run  `git config blame.ignoreRevsFile .ignoreRevsFile`  once there is a `.ignoreRevsFile` file. 

3. Reformat sections of OpenSC that do not change much and don't have any outstanding PRs. 

4. Schedule reformats for the rest of OpenSC based on developers schedules, or as part of a larger PR with the reformat as its own commit. 

5. Automate checking of PRs, so they meet the new format formats.

**Comment by dengert** [comment]
@Jakuje can you also add #2185 and add commit to .ignoreRevFiles

**Comment by Jakuje** [comment]
> @Jakuje can you also add #2185 and add commit to .ignoreRevFiles

Added. Reviewed and it looks fine. If there will be no objections, I will merge it as soon there will be some CI results.

**Comment by dengert** [comment]
Looks like windows has a problem with every program that includes `unistd.h`
looks like change to `opensc/types.h`  in 
5ae9882e1bdda7f0cb1bf647480bdd7f81bb2609 needs `#ifdef HAVE_UNISTD_H`

**Comment by Jakuje** [comment]
> Looks like windows has a problem with every program that includes unistd.h
> looks like change to opensc/types.h in
> 5ae9882 needs #ifdef HAVE_UNISTD_H

Thanks. Fixed in current version. As well as I updated the commit from #2185 after merging.

@frankmorgner whats up with the OSX builds. They keep failing in pull requests( but work in master). Is it something we can fix:
```
/Users/travis/build/OpenSC/OpenSC/OpenSC.tokend/build/security_cdsa_utilities.framework/Headers/cssmdata.h:510:29: warning: implicit conversion loses integer precision: 'unsigned long' to 'uint32' (aka 'unsigned int') [-Wshorten-64-to-32]

                uint32 newLength = Length + data.Length;

                       ~~~~~~~~~   ~~~~~~~^~~~~~~~~~~~~

** INSTALL FAILED **

The following build commands failed:

	CompileC build/Tokend.build/OpenSC.build/Objects-normal/x86_64/opensc.o OpenSC/opensc.cpp normal x86_64 c++ com.apple.compilers.llvm.clang.1_0.compiler

(1 failure)

The command "if [ "${DO_COVERITY_SCAN}" != "yes" ]; then if [ "${TRAVIS_OS_NAME}" = "osx" ]; then ./MacOSX/build; else make -j 4; fi; fi" exited with 65.
```

**Comment by frankmorgner** [comment]
Most of the TokenD boilerplate was originally taken from Apple. I won't do anything to fix those tons of warnings with it, because it's deprecated by CryptoTokenKit. However, the actual error is different from what you've quoted:
```
In file included from /Users/travis/build/OpenSC/OpenSC/OpenSC.tokend/OpenSC/OpenSCToken.h:35:

/Users/travis/build/OpenSC/OpenSC/OpenSC.tokend/../src/libopensc/opensc.h:40:10: fatal error: 'config.h' file not found

#include "config.h"
```
I think that should be solved by modifiying `HEADER_SEARCH_PATHS` to add the location of `config.h`. I'll try to propose a diff when I have a mac at hand...

**Comment by frankmorgner** [comment]
@Jakuje do you think it's ready now? should I review once more?

There are still some minor code changes required to let clang-format pass initially...

**Comment by Jakuje** [comment]
> @Jakuje do you think it's ready now? should I review once more?
> 
> There are still some minor code changes required to let clang-format pass initially...

I finally rebased the changes on top of current master + fixed some of the formatting reported by the tool since 0.22.0.

I also added the `AfterFunction` brace attachment, which is very common in the new code, but uncommon in the old code. I brought this up already in https://github.com/OpenSC/OpenSC/pull/2017#issuecomment-744433639 without any follow up. I already documented the "new" way in `CONTRIBUTE.md`.

There are couple of reflows left when I compare this against 0.22.0, but the next big change is how macros are rendered with the continuation to the next line. The current configuration prefers the following format, which I am still not convinced is the best one (as most of the new code does not use it), but I do not have any better suggestion:
```
diff --git a/src/libopensc/opensc.h b/src/libopensc/opensc.h
index 5b26a809..da830d3c 100644
--- a/src/libopensc/opensc.h
+++ b/src/libopensc/opensc.h
@@ -1540,10 +1540,11 @@ int sc_base64_decode(const char *in, u8 *out, size_t outlen);
 void sc_mem_clear(void *ptr, size_t len);
 void *sc_mem_secure_alloc(size_t len);
 void sc_mem_secure_free(void *ptr, size_t len);
-#define sc_mem_secure_clear_free(ptr, len) do { \
-	sc_mem_clear(ptr, len); \
-	sc_mem_secure_free(ptr, len); \
-} while (0);
+#define sc_mem_secure_clear_free(ptr, len)                                                                   \
+	do {                                                                                                 \
+		sc_mem_clear(ptr, len);                                                                      \
+		sc_mem_secure_free(ptr, len);                                                                \
+	} while (0);
 int sc_mem_reverse(unsigned char *buf, size_t len);
 ```

I would still like the others to check the config and changes whether they work for you. Certainly before merging the rewrites and rewriting all the rest of the code. As previously mentioned, I am fine with having this as imperfect CI action until we settle on final formatting.

**Comment by Jakuje** [comment]
Sigh ... it looks like the action we use runs in some outdated container that does not support
```
YAML:25:9: error: unknown enumerated scalar
UseTab: AlignWithSpaces
        ^~~~~~~~~~~~~~~
Error reading /github/workspace/.clang-format: Invalid argument
error: `clang-format -lines=23:25 src/common/libpkcs11.h` failed
```
With https://github.com/yshui/git-clang-format-lint/pull/3 it looks like it works fine

**Comment by frankmorgner** [comment]
What's the status of this topic? We can surely debate later on the actual style, but we should fix the CI error if we want to accept the currently suggested style template.

**Comment by Jakuje** [comment]
I would love to get this merged. I just spent some time rebasing this branch and applying formatting changes since 0.22.0 release (just to get some snapshot). They are available in this PR (not final though). But I am not completely happy with the formatter or configuration yet. There are still corner cases where I would like to keep the original format without completely turning the formatter off:

 * The closing brace of the structures definition. For example, the clang now proposes the following changes:
```
-       {NULL, NULL, NULL, 0, 0, NULL}
-};
+       {NULL, NULL, NULL, 0, 0, NULL}};
```
 * I liked to keep the return type of a function on a separate line as it makes the long lines shorter, especially with many arguments:
```
-static int 
-default_get_cvc_ca_cert(sc_card_t *card, u8 **cert, size_t *length)
+static int default_get_cvc_ca_cert(sc_card_t *card, u8 **cert, size_t *length)
```
 * For complicated conditions, it tries to squeeze more of them  on a single line, which I consider less readable than keeping them on separate lines:
```
-       if (!EVP_DigestInit_ex(ctx, digest, NULL) ||
-           !EVP_DigestUpdate(ctx, input, length)) {
+       if (!EVP_DigestInit_ex(ctx, digest, NULL) || !EVP_DigestUpdate(ctx, input, length)) {
```
 * It does not like the indented includes. But I think this is a minor issue:
```
-# include <openssl/core_names.h>
+#include <openssl/core_names.h>
```
 * It reflows the function arguments to fill the line length, but I would like to keep the pairs (value & length) together:
```
-static int cwa_prepare_external_auth(sc_card_t *card, EVP_PKEY *icc_pubkey, EVP_PKEY *ifd_privkey,
-                                     u8 *sig, size_t sig_len)
+static int cwa_prepare_external_auth(sc_card_t *card, EVP_PKEY *icc_pubkey, EVP_PKEY *ifd_privkey, u8 *sig,
+                                     size_t sig_len)
```
 * The multiline macros -- we did not touch them so far much, but I am not sure if I like this. I agree that it is readable, but I feel like there is too many needless whitespace and even more processing needed with every change you do to this macro. Moreover most of our code is not formed like this so we might need to figure out better format configuration
```
-#define sc_debug_hex(ctx, level, label, data, len) \
-    _sc_debug_hex(ctx, level, FILENAME, __LINE__, __FUNCTION__, label, data, len)
+#define sc_debug_hex(ctx, level, label, data, len)                                                           \
+       _sc_debug_hex(ctx, level, FILENAME, __LINE__, __FUNCTION__, label, data, len)
```

Otherwise, it finds many useful issues which could save us a lot of time reviewing the PRs, verifying there is no stray whitespace, indentation makes sense, lines are not too long and much more. So I would love to have this around, but I would like to have it to only warn (afaik this is not possible in GH) if it finds the issue and still keep a common sense reviewing the code.

So I would be happy for feedback to the current formatter configuration, to the issues outlined above, other concerns or suggestions for improvements.

Just for the info, the GH action running clang format was already fixed so the issue I mentioned in Septemer is no longer an issue.

**Comment by frankmorgner** [comment]
In the workflow we can limit the formatting complaints to pull requests only. Would it possible to restrict the formatting complaints further to the changes proposed by a pull request? If so, we can surely live with the some pieces of the code in master not complying with the style suggested by clang-format and postpone this for later.

**Comment by Jakuje** [comment]
> In the workflow we can limit the formatting complaints to pull requests only.

Thats currently done.

> Would it possible to restrict the formatting complaints further to the changes proposed by a pull request?

This is also the current behavior.

> If so, we can surely live with the some pieces of the code in master not complying with the style suggested by clang-format and postpone this for later.

That was my initial proposal, but I think @dengert defended we need to convert the whole codebase first with the `.ignoreRevsFile` file for cleaner blame/history. I agree with this request as the suggestions to wrongly formatted code are sometimes very confusing.

I did progress with some questions posted above while rebasing and resolving the conflicts.

>  * I liked to keep the return type of a function on a separate line as it makes the long lines shorter, especially with many arguments:
> ```
> -static int 
> -default_get_cvc_ca_cert(sc_card_t *card, u8 **cert, size_t *length)
> +static int default_get_cvc_ca_cert(sc_card_t *card, u8 **cert, size_t *length)
> ```

It turns out this is handled with `AlwaysBreakAfterReturnType: AllDefinitions` (the declarations are usually not broken), so this would be another change I would propose.

> * The multiline macros -- we did not touch them so far much, but I am not sure if I like this. I agree that it is readable, but I feel like there is too many needless whitespace and even more processing needed with every change you do to this macro. Moreover most of our code is not formed like this so we might need to figure out better format configuration
> 
> 
> ```
> -#define sc_debug_hex(ctx, level, label, data, len) \
> -    _sc_debug_hex(ctx, level, FILENAME, __LINE__, __FUNCTION__, label, data, len)
> +#define sc_debug_hex(ctx, level, label, data, len)                                                           \
> +       _sc_debug_hex(ctx, level, FILENAME, __LINE__, __FUNCTION__, label, data, len)
> ```

This is controlled with `AlignEscapedNewlines: DontAlign`. I would also propose to use this to avoid reformatting all macros we have. It does not match our macros completely, but more closer than the previous version.

All changes squashed in the top commit (rebased on current master)

**Comment by frankmorgner** [comment]
>> ... we can surely live with the some pieces of the code in master not complying with the style suggested by clang-format and postpone this for later.

>That was my initial proposal, but I think @dengert defended we need to convert the whole codebase first with the `.ignoreRevsFile` file for cleaner blame/history. I agree with this request as the suggestions to wrongly formatted code are sometimes very confusing.

Great to hear, that it's already that far. Considering that the style definition is in the queue for so long, we should go ahead and merge this PR in its current version.

For sure, we will continue to tweak the configuration options in the near and far future once we stumble upon a change that we don't like. That's why I think making a big formatting commit will almost surely be unsatisfying as it introduces unwanted changes and it will cause a lot of work porting people's local changes to comply with the reformatted code. Incremental changes can be handled more easily.

@dengert , would you be OK with that proposal?

Regarding the specific options that you suggested, Jakub, it's the right approach to keep the current implicit style rules intact. The initial proposal looks good enough, the refinement can be discussed in smaller pull requests.

**Comment by dengert** [comment]
I don't want to stand in the way of this PR.

The only concern I have deals with in active PRs, or other works-in-progress developers are working on. They will have to do a rebase some time. If we can deal with the massive changes OpenSSL we can deal with reformats of PRs. 

I would suggest that you pick a date and time so developers can rebase their changes just before this PR is committed, then they can rebase just after this PR is committed. So in effect their rebase would consist of format only changes and included in the `.ignoreRevFile`  Actually a developer could rebase on the commit before this PR then rebase on the commit after this PR at any time.  So you don't have to really pick a date, just let people know what they need to do.

Distro modifications that other will have to deal with in the next release.

**Comment by Jakuje** [comment]
Note, that I do not insist on merging this whole PR as it rewrites a lot of code and not always to the strict formatting enforced by the clang. I mostly submitted it here for review if the formatting makes sense for me as I wrote it and for anyone who would like to see some larger chunks of the code reformatted with the proposed rules. But I would like to keep at least the `clang-format on/off`, and I do not mind much about git history in the testing/fuzzing code, so I would like to apply changes there too.

I will rebase the PR to separate the changes a bit better later today.

> * For complicated conditions, it tries to squeeze more of them  on a single line, which I consider less readable than keeping them on separate lines:
> 
> 
> ```
> -       if (!EVP_DigestInit_ex(ctx, digest, NULL) ||
> -           !EVP_DigestUpdate(ctx, input, length)) {
> +       if (!EVP_DigestInit_ex(ctx, digest, NULL) || !EVP_DigestUpdate(ctx, input, length)) {
> ```

This is addressed with `ColumnLimit: 0`, which prevents reflowing intentionally short lines. But I think we need a baseline for maximum length defined somewhere so this is something I would consider using this only in the mass-reformat if we would get to that. Right now, there are very long lines or lines broken in 80 characters throughout the code which need some treatment.

**Comment by frankmorgner** [comment]
> @OpenSC/maintainers question/comments/concerns again? I would suggest the end of March to merge at least first commits by that date. I keep the formatting changes in for now for preview purposes for now.

Oh, I totally missed that it also reformats all the files 🙈. My mind was still with #1641 which only introduced the definition and I thought this PR was extending this to the changes of a PR. I'll try to read through the actual changes of the files today...

File: src/libopensc/pkcs15-iasecc.c
Comment (frankmorgner) [inline src/libopensc/pkcs15-iasecc.c:113]: shouldn't this be `const struct sc_context *ctx`?

File: src/scconf/parse.c
Comment (frankmorgner) [inline src/scconf/parse.c:329]: shouldn't the `break` go on an extra line?

File: src/tools/pkcs11-tool.c
Comment (frankmorgner) [inline src/tools/pkcs11-tool.c:5586]: Here you have an example where the formatting actually makes the code harder to read... The block encapsulated in `{}` is suddenly aligned identically as the conditions from the `if` clauses. So let's don't mix tabs/spaces to align the breaks with the opening brackets as argued above

**Review by frankmorgner — CHANGES_REQUESTED** [review]
There are some header changes, which I didn't expect here, but they should not do any harm. So I think it's OK to leave them as they are in this PR.

One thing that needs to change, I think, is the mix of tabs and spaces. This is often done to align the content with an opening bracket after a line break, for example. But I think this alignment is not very useful, for example when people have a different taste on how much space a tab uses in their viewer. I suggest to simply use additional indents (i.e. tabs) after a line break.

I personally like to not alight a `case` with `switch` but to rather use an additional indent, but I don't insist on this one.

File: src/libopensc/pkcs15-iasecc.c
Comment (Jakuje) [reply src/libopensc/pkcs15-iasecc.c:113]: right. Missed that when rewriting too many changes. I will fix it.

File: src/libopensc/card-atrust-acos.c
Comment (Jakuje) [reply src/libopensc/card-atrust-acos.c:724]: Thats the intention to make it aligned regardless the tab length.

But you are right that this is not much used in the code. I think what you describe woudl be written as
```
AlignAfterOpenBracket: DontAlign
ContinuationIndentWidth: 2
UseTab: ForContinuationAndIndentation
```
but I am afraid it will make much longer lines for the `if` conditions. But let me try.

File: src/libopensc/card-epass2003.c
Comment (Jakuje) [reply src/libopensc/card-epass2003.c:2401]: This is also intentional. These are only spaces to keep the alignment regardless the tab width.

File: src/scconf/parse.c
Comment (Jakuje) [reply src/scconf/parse.c:329]: probably yes.Or inside of the braces.

File: src/tools/pkcs11-tool.c
Comment (Jakuje) [reply src/tools/pkcs11-tool.c:5586]: This is true for one configuration where you have tab width=4 (because the length of "if (" is 4). For all the other cases, the block is aligned differently than the condition.

File: src/libopensc/card-atrust-acos.c
Comment (Jakuje) [reply src/libopensc/card-atrust-acos.c]: I am playing with this a bit with the following configuration:
```
+TabWidth: 4
+IndentWidth: 4
 ContinuationIndentWidth: 8
-UseTab: AlignWithSpaces
+AlignAfterOpenBracket: DontAlign
+UseTab: ForContinuationAndIndentation
```
And the hardcoded `ContinuationIndentWidth` has a result of all the structs and tables definition are indented with two tabs, which is something I am not very happy with. Also if we have a function call with many arguments in the `if` condition, the overflowing arguments are indented with one more tab, which can go too much right (again depending on the tab width of the editor). The same for broken srings in the function arguments.

One solution for this problem might be not allowing function calls in the `if` conditions, because this makes arguably hard to read and debug code, long lines and excess indentation. We have similar guides in libssh: https://gitlab.com/libssh/libssh-mirror/-/blob/master/CONTRIBUTING.md#make-use-of-helper-variables but again, this does not fit the "current use in opensc" -- just a suggestion.

What I found confusing in this formatting configuration:

 * The long lines sometimes result in the following reformat, which I find ugly and I did not find a way to prevent it:
```
-idprime_read_binary(sc_card_t *card, unsigned int offset, unsigned char *buf, size_t count,
-               unsigned long flags)
+idprime_read_binary(
+               sc_card_t *card, unsigned int offset, unsigned char *buf, size_t count, unsigned long flags)
```
 * The structure/tables definitions are indented with 2 tabs instead of one:
```
        const u8 *aids[] = {
-               (const u8 *)"\xa0\x00\x00\x02\x47\x10\x01",
+                       (const u8 *)"\xa0\x00\x00\x02\x47\x10\x01",
```
 * The strings are still aligned with some spaces instead of tabs for some reason:
```
                                parse_error(cur, "Template insane: file-path length should not be "
-                                               "less than 2 bytes");
+                                                "less than 2 bytes");
```
The most crazy stuff is avoided with the `ColumnLimit: 0`, but in some other cases, it still tries to indent with spaces and I have no idea why in some cases where long function calls with a lot of arguments is used.

I just pushed the updated commits. Would be glad for another look if you would be more happy with them then I am.

**Review by frankmorgner — CHANGES_REQUESTED** [review]
I personally would prefer 
[BinaryOperatorStyle](`https://clang.llvm.org/docs/ClangFormatStyleOptions.html#breakbeforebinaryoperators`), because in if-conditions I like to have binary operators in front to make it more visible

But I can also live with the current formatting.

AFAICT, there are two review comments that need be resolved. Is there anything else to do to get this merged (besides fixing the merge conflicts)?

File: src/scconf/parse.c
Comment (Jakuje) [reply src/scconf/parse.c:329]: I rather rewrote this case not to use the block as this did not look like changing anything for me. But added anyway.

File: src/libopensc/opensc.h
Comment (Jakuje) [reply src/libopensc/opensc.h:40]: Done.

File: src/libopensc/card-atrust-acos.c
Comment (Jakuje) [reply src/libopensc/card-atrust-acos.c]: @frankmorgner there are some unresolved issues of the current configuration. Additionally, with the current configuration, I see some issues with alignment of continuation of function calls in `if(` conditions, such as:
```
@@ -3500,9 +3500,9 @@ wrap_key(CK_SESSION_HANDLE session)
 		util_fatal("Secret key (to be wrapped) not found");
 
 	if (!find_object(session, CKO_PUBLIC_KEY, &hWrappingKey, opt_object_id_len ? opt_object_id : NULL,
-			opt_object_id_len, 0))
+			    opt_object_id_len, 0))
 		if (!find_object(session, CKO_SECRET_KEY, &hWrappingKey,
-				opt_object_id_len ? opt_object_id : NULL, opt_object_id_len, 0))
+				    opt_object_id_len ? opt_object_id : NULL, opt_object_id_len, 0))
 			util_fatal("Public/secret key (wrapping key) not found");
 
 	rv = p11->C_WrapKey(session, &mechanism, hWrappingKey, hkey, pWrappedKey, &pulWrappedKeyLen);
```
I find this totally non-intuitive, as the three (or 4) tabs are followed by 4 spaces (that probably account for the `if (` string?). This all with `UseTab: Always` (this is not going away even with `UseTab: ForContinuationAndIndentation`). FYI I am using clang 14 now.

Comment (frankmorgner) [reply src/libopensc/card-atrust-acos.c]: I don't know exactly why, but this will actually result in tabs instead of spaces:
```
ContinuationIndentWidth: 20
AlignAfterOpenBracket: DontAlign
```

Comment (frankmorgner) [reply src/libopensc/card-atrust-acos.c]: hmm, the above setting will then result in a mix of tabs and spaces when used in a case like this:
```

int i =         //  VeryVeryVeryVeryVeryLongComment
  longFunction( // Again a long comment
    arg);
```
So this is not yet solved

Comment (frankmorgner) [reply src/libopensc/card-atrust-acos.c]: In my tests, this here seems to do what you want:
```
TabWidth: 4
IndentWidth: 4
ContinuationIndentWidth: 8
AlignAfterOpenBracket: DontAlign
```

Comment (Jakuje) [reply src/libopensc/card-atrust-acos.c]: Thanks for having a look. But with this configuration, I did not notice any significant change. It looks like it resolved the issue with the `if`, but there are very similar cases with `else if` case, which it still tries to format with tabs + spaces, see the example below:
```
@@ -1380,9 +1380,9 @@ int main(int argc, char * argv[])
 			find_object_flags(session, mf_flags, &object,
 					opt_object_id_len ? opt_object_id : NULL, opt_object_id_len, 0);
 		} else if (!find_object(session, CKO_PUBLIC_KEY, &object,
-					opt_object_id_len ? opt_object_id : NULL, opt_object_id_len, 0) &&
-				!find_object(session, CKO_CERTIFICATE, &object,
-					opt_object_id_len ? opt_object_id : NULL, opt_object_id_len, 0))
+						   opt_object_id_len ? opt_object_id : NULL, opt_object_id_len, 0) &&
+				   !find_object(session, CKO_CERTIFICATE, &object,
+						   opt_object_id_len ? opt_object_id : NULL, opt_object_id_len, 0))
 			util_fatal("Public key nor certificate not found");
 	}
 
```
I also still get all the structure initializers indented with two tabs which I find as a huge waste of horizontal space (see the results of the current CI run). But otherwise, I find it mostly usable.

One more thing to consider is alignment of `type variable comment` initializers. This is used quite a lot in the code, but not much in the new one as I do not find it very useful. Example:
```
 	unsigned char sk_enc[16];	/* encrypt session key */
 	unsigned char sk_mac[16];	/* mac session key */
 	unsigned char icv_mac[16];	/* instruction counter vector(for sm) */
-	unsigned char bFipsCertification;	/* fips mode Alg */
+	unsigned char bFipsCertification; /* fips mode Alg */
```
The formatter tries to get rid of the additional whitespace, causing large diffs too. In this case, this is aligned with tabs so it looks completely different in different editors. What would be your take on these changes?

Comment (frankmorgner) [reply src/libopensc/card-atrust-acos.c]:
```
UseTab: AlignWithSpaces
```
fixes this:
>Use tabs for line continuation and indentation, and spaces for alignment

I think we can agree on the fact that clang-format will never provide a perfect result. And the above example of mixing tabs and spaces for indentation sounds like a bug in clang-format.

In my tests I've noticed that my preferred option `AlignAfterOpenBracket: AlwaysBreak` together with `ColumnLimit: 0` results in a removal of intentionally placed line breaks. To fix this, I *could* argue that we should not hold some changes back to reduce the number of changes in this pull request (i.e. the change to `ColumnLimit: 110`). But as said before, I think I can live with the imperfections we are currently seeing and I'd like to move on with the format in its current state.

Comment (Jakuje) [reply src/libopensc/card-atrust-acos.c]: I reformated the structures to be indented with 2 tabs. Now, some of them are also formatted with `AlignArrayOfStructures: Left`, but this option keeps crashing my clang so I left it out yet.
There are still some weirdnesses (and the commit series is mess now) so I would propose to merge only the documentation + .clang-format + gh action (I will split it) and the rest of reformatting only after some better review. Its too large.

File: src/scconf/parse.c
Comment (Jakuje) [reply src/scconf/parse.c]: > please try setting (IndentCaseBlocks)[https://clang.llvm.org/docs/ClangFormatStyleOptions.html#indentcaseblocks] to `true`

Looked again at this option, and I think most of the existing code is using the `false` value. Rewriting everything to match `true` would again lead to a lot of useless indentation. What do you think?

File: src/libopensc/card-atrust-acos.c
Comment (Jakuje) [reply src/libopensc/card-atrust-acos.c]: I pused couple of more changes built with clang 15, where this option seems to be working. I did some changes that looked like made sense to me, but ignored some where we did not completely agreed or they were only reflowing ling breaks to longer lines.

With that I am still not completely confident the clang does the right thing in the following cases:
 * the `else if` continuation indentation is weird (as already mentioned)
 * multiline the ternary operator is still indented with spaces
 * with `AlignArrayOfStructures`, if there is many spaces used to align the structure members, some of the spaces are replaced with tabs, which breaks the indentation (at least in the diff view, which probably does not follow our tab width -- but if it looks different with different settings, its kind of useless)
 * aligning of type + variable + values + comments in the variable declaration -- half of the code does it and the other does not do that. What would be your take?

Comment (frankmorgner) [reply src/libopensc/card-atrust-acos.c]: As said before, it doesn't need to be perfect. For now, it is nice to
- have at least some definition written down and
- automate the process of highlighting formatting problems in PRs

😮‍💨

Comment (Jakuje) [reply src/libopensc/card-atrust-acos.c]: Yeah. Thats what I tried to put in contributing.md Please, check if it covers all what is needed.

**Comment by popovec** [comment]
Apparently, this PR is not in a condition to be merged as it is, but I would ask if at least the single .clang-format file from this package could not be merged into the master. Thank you.

**Comment by Jakuje** [comment]
> Apparently, this PR is not in a condition to be merged as it is, but I would ask if at least the single .clang-format file from this package could not be merged into the master. Thank you.

Yeah. I spent already too much time reformatting the code, while trying to understand some of the options and their behavior. I moved the non-problematic changes into separate PR #2977. Some less disruptive changes on top of them are in https://github.com/Jakuje/OpenSC/tree/clang-format for preview how it tries to format some larger code changes (the leftovers marked as clang format now are mostly the changes that I do not like or missed).

**Pull request closed without merging**
